### PR TITLE
Export NaNs in logits to scheduler_stats if output is corrupted

### LINKED ISSUE #17123: [Feature]: Automatically detect numerical issues
### 🚀 The feature, motivation and pitch

Models such as Gemma-3 and GLM-4 may encounter numerical instability when `float16` dtype is used. Despite a warning message `Casting bfloat16 to float16` being printed, users can still get confused when the model returns empty or nonsense outputs.

Examples:
- https://github.com/vllm-project/vllm/issues/16489
- https://github.com/vllm-project/vllm/pull/16618#issuecomment-2814399522

It would be great if we could automatically detect numerical issues while running the model. We should at least check for this during startup. Inference-time checking should be optional since it harms the performance.

### Alternatives

Hardcode specific models to not allow them to be run in float16, like `plamo2`:

https://github.com/vllm-project/vllm/blob/4115f19958d8b3628606d78355c277b328f011e1/vllm/config.py#L2834

However, this has to be done manually.

### Additional context

cc @youkaichao 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by vladmihailescu**
Signed-off-by: Vlad Mihailescu vladmihailescu@meta.com

Summary:
Report nan in logits in scheduler_stats. This can be used later exported as Phrometeus counter but for now this is required so we can export it in our internal counter infra.

This counter is used to identify bad hosts or bad GPUs which cause NaNs in logits.

 It's a common metric we expose.

Reviewed By: Adolfo-Karim

Differential Revision: D75423285

```
NO ENV VAR (NaN counting off):
I0618 22:24:27.611000 180881 reporter.py:195] Part 1: High-level performance metrics
Ran 404/404 requests in 56.56s
Success rate:        100.00%
QPS:                 7.14
Avg latency:         11.361s
Avg TTFT (client):   8245.31ms
P50 TTFT (client):   8262.13ms
P99 TTFT (client):   8283.14ms
Avg TTIT (client):   311.60ms
P50 TTIT (client):   323.78ms
P99 TTIT (client):   334.87ms
Avg TTFT (server):   6748.17ms
Avg TTIT (server):   357.83ms
Avg prefill len:     2587.54 tokens
P50 prefill len:     2587.00 tokens
P99 prefill len:     2596.00 tokens
Avg decode len:      10.00 tokens
P50 decode len:      10.00 tokens
P99 decode len:      10.00 tokens
I0618 22:22:13.811000 141492 reporter.py:195] Part 1: High-level performance metrics
Ran 404/404 requests in 56.30s
Success rate:        100.00%
QPS:                 7.18
Avg latency:         11.312s
Avg TTFT (client):   8209.60ms
P50 TTFT (client):   8214.03ms
P99 TTFT (client):   8273.95ms
Avg TTIT (client):   310.22ms
P50 TTIT (client):   322.10ms
P99 TTIT (client):   325.09ms
Avg TTFT (server):   6639.55ms
Avg TTIT (server):   355.36ms
Avg prefill len:     2587.49 tokens
P50 prefill len:     2587.00 tokens
P99 prefill len:     2595.00 tokens
Avg decode len:      10.00 tokens
P50 decode len:      10.00 tokens
P99 decode len:      10.00 tokens
```

```
WITH ENVVAR (NaN counting on):
I0618 22:44:09.103000 508486 reporter.py:195] Part 1: High-level performance metrics
Ran 404/404 requests in 56.18s
Success rate:        100.00%
QPS:                 7.19
Avg latency:         11.291s
Avg TTFT (client):   8193.95ms
P50 TTFT (client):   8208.52ms
P99 TTFT (client):   8232.08ms
Avg TTIT (client):   309.75ms
P50 TTIT (client):   321.92ms
P99 TTIT (client):   333.03ms
Avg TTFT (server):   8574.45ms
Avg TTIT (server):   355.91ms
Avg prefill len:     2587.90 tokens
P50 prefill len:     2588.00 tokens
P99 prefill len:     2596.00 tokens
Avg decode len:      10.00 tokens
P50 decode len:      10.00 tokens
P99 decode len:      10.00 tokens
I0618 22:47:21.294000 566871 reporter.py:195] Part 1: High-level performance metrics
Ran 404/404 requests in 57.23s
Success rate:        100.00%
QPS:                 7.06
Avg latency:         11.532s
Avg TTFT (client):   8365.60ms
P50 TTFT (client):   8229.34ms
P99 TTFT (client):   9156.80ms
Avg TTIT (client):   316.59ms
P50 TTIT (client):   322.71ms
P99 TTIT (client):   414.20ms
Avg TTFT (server):   6794.25ms
Avg TTIT (server):   358.08ms
Avg prefill len:     2587.78 tokens
P50 prefill len:     2587.00 tokens
P99 prefill len:     2597.00 tokens
Avg decode len:      10.00 tokens
P50 decode len:      10.00 tokens
P99 decode len:      10.00 tokens
```

And unit test

```
pytest -s -v tests/v1/worker/test_gpu_model_runner.py
```
```
tests/v1/worker/test_gpu_model_runner.py::test_get_nans_in_logits INFO 06-19 02:06:26 [config.py:1444] Using max model len 2048
WARNING 06-19 02:06:26 [config.py:4758] Current vLLM config is not set.
WARNING 06-19 02:06:26 [config.py:4758] Current vLLM config is not set.
WARNING 06-19 02:06:26 [topk_topp_sampler.py:59] FlashInfer is not available. Falling back to the PyTorch-native implementation of top-p & top-k sampling. For the best performance, please install FlashInfer.
PASSED
```

FIX https://github.com/vllm-project/vllm/issues/17123

### CODE REVIEW

**Comment by simon-mo** [comment]
Keep in mind that the logits needs to be serialized from model runner back to the scheduler via RPC/collectives. So doing the counting in model runner will be better.

**Comment by markmc** [comment]
Some overlap with #18765 ... except I'm not sure this NaN case results in the request explicitly failing?

**Review by yeqcharlotte — CHANGES_REQUESTED** [review]
let's make sure we have tests for the new metrics. 

nan can happen both due to hardware failures or when we mess up quantization's calibration. i agree with @markmc we should check the overlap with failed requests. 

could you simulate some nans by mocking some kernel outputs in the middle of generation and see if we were able to see request status and finish reasons getting populated correctly.

**Comment by yeqcharlotte** [comment]
lgtm @vladmihailescu could you update the test plan? any data to show this is of low load overhead? possible to add a unit test to ensure it works?

cc: @simon-mo @WoosukKwon @njhill @houseroad could you help to take a look?

**Review by houseroad — COMMENTED** [review]
Can we have a environment var to control this? I am a bit concerned that it may cause some overhead. Or we can do some benchmark to confirm the overhead.

**Comment by vladmihailescu** [comment]
@yeqcharlotte @houseroad I added an unit test and gated the computation behind an env var which is off by default. I ran few benchmarks with loadgen and there's no significant throughput/latency difference at high concurrency but I didn't look into traces. Are you suggesting looking at gpu traces to check the overhead or something else? If the loadtest is enough I think we can also do shadow internally after this lands.

**Review by houseroad — APPROVED** [review]
Looks good to me. Thanks for adding this metrics.

**Comment by vladmihailescu** [comment]
V1 test timed out. Rebasing after the fix (https://github.com/vllm-project/vllm/pull/19872)

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -1826,6 +1831,25 @@ def _get_prompt_logprobs_dict(
 
         return prompt_logprobs_dict
 
+    def _get_nans_in_logits(
+        self,
+        logits: torch.Tensor,
```
Comment (yeqcharlotte) [inline vllm/v1/worker/gpu_model_runner.py]: is `logits` optionally None here or not?

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -1826,6 +1831,25 @@ def _get_prompt_logprobs_dict(
...
+            num_nans_for_index = None
+            if logits is not None:
+                num_nans_for_index = logits.isnan().sum(dim=-1).cpu().numpy()
+            for req_id in self.input_batch.req_ids:
+                req_index = self.input_batch.req_id_to_index[req_id]
+                num_nans_in_logits[req_id] = (
+                    int(num_nans_for_index[req_index])
+                    if logits is not None and num_nans_for_index is not None
+                    and req_index < logits.shape[0] else 0)
+            return num_nans_in_logits
```
Comment (yeqcharlotte) [inline vllm/v1/worker/gpu_model_runner.py]: nit: you got 2 `if logits is not none` check here and one in the loop. for better readability it might make sense to do the following:

```
if logits is None:
    return {req_id: 0 for ...}
num_nans_for_index = logits.isnan().sum(dim=-1).cpu().numpy()    
# count nans
```

**Review by yeqcharlotte — APPROVED** [review]
thanks for adding this! leaving some small nits only.

**Comment by vladmihailescu** [comment]
Addressed nits
